### PR TITLE
feat: Javascript Plugin API (Custom panels, column menu items with JS actions)

### PR DESCRIPTION
**Pull request by hydrosquall**
## Motivation

- Allow plugins that add data visualizations [`datasette-vega`](https://github.com/simonw/datasette-vega), [`datasette-leaflet`](https://github.com/simonw/datasette-leaflet), and [`datasette-nteract-data-explorer`](https://github.com/hydrosquall/datasette-nteract-data-explorer) to co-exist safely
- Standardize APIs / hooks to ease development for new JS plugin developers (better compat with datasette-lite) through standardized DOM selectors, methods for extending the existing Table UI. This has come up as a feature request several times (see research notes for examples)
- Discussion w/ @simonw about a general-purpose Datasette JS API

## Changes

Summary: Provide 2 new surface areas for Datasette JS plugin developers

1. Custom column header items: <https://a.cl.ly/Kou97wJr>
2. Basic "panels" controlled by buttons: <https://a.cl.ly/rRugWobd>

### User Facing Changes

- Allow creating menu items under table header that triggers JS (instead of opening hrefs per the existing [menu_link](https://docs.datasette.io/en/stable/plugin_hooks.html#menu-links-datasette-actor-request) hook). Items can respond to any column metadata provided by the column header (e.g. label). The proof of concept plugins log data to the console, or copy the column name to clipboard.
- Allow plugins to register UI elements in a panel controller. The parent component handles switching the visibility of active plugins.
  - Because native button elements are used, the panel is keyboard-accessible - use tab / shift-tab to cycle through tab options, and `enter` to select.
  - There's room to improve the styling, but the focus of this PR is on the API rather than the UX.

### (plugin) Developer Facing Changes

- Dispatch a `datasette_init` [CustomEvent](https://developer.mozilla.org/en-US/docs/Web/API/CustomEvent/CustomEvent) when the `datasetteManager` is finished loading.
- Provide `manager.registerPlugin` API for adding new functionality that coordinates with Datasette lifecycle events.
- Provide a `manager.selectors` map of DOM elements that users may want to hook into.
  - Updated `table.js` to use refer to these to motivating keeping things in sync
- Allow plugins to register themselves with 2 hooks:
  - `makeColumnActions`: Add items to menu in table column headers. Users can provide a `label`, and either `href` or `onClick` with full access to the metadata for the clicked column (name, type, misc. booleans)
  - `makeAboveTablePanelConfigs`: Add items to the panel. Each panel has a unique ID (namespaced within that plugin), a render function, and a string label.

See [this file](https://github.com/simonw/datasette/blob/2d92b9328022d86505261bcdac419b6ed9cb2236/datasette/static/table-example-plugins.js) for example plugin usage.

###  Core Developer Facing Changes

- Modified `table.js` to make use of the `datasetteManager` API.
- Added example plugins to the `demos/plugins` folder, and stored the test js in the `statics/` folder

## Testing

For Datasette plugin developers, please see the [alpha-level documentation](https://github.com/simonw/datasette/pull/2052#issuecomment-1510423051) .

To run the examples:

```bash
datasette serve fixtures.db  --plugins-dir=demos/plugins/
```

Open local server: `http://127.0.0.1:8001/fixtures/facetable`

Open to all feedback on this PR, from API design to variable naming, to what additional hooks might be useful for the future.

My focus was more on the general shape of the API for developers, rather than on the UX of the test plugins.

## Design notes

- The manager tab panel could be a separate plugin if the implementation is too custom.
- The `makeColumnHeaderItems` benefits from hooking into the logic of `table.js`
- I wanted to offer this to the Datasette core, since the `datasette-manager` would be more powerful if it were connected to lifecycle and JS events that are part of the existing table.js.
- Non-goals:
  - Dependency management (for now) - there's no "build" step, we don't know when new plugins will be added. While there are some valid use cases (for example, allow multiple plugins to wait for a global leaflet object to be loaded), I don't see enough use-cases to justify doing this yet.
  - Enabling single-page-app features - for now, most datasette actions lead to a new page being loaded. SPA development offers some benefits (no page jumping after clicking on a link), but also complexity that doesn't need to be in the core Datasette project.

## Research Notes

- Relocated to a [comment](https://github.com/simonw/datasette/pull/2052#issuecomment-1510423215), as this isn't required to review when evaluating the plugin. Including it just for those who are curious.

### CODE REVIEW

File: datasette/static/table-example-plugins.js [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+});
+
+/**
+ * Examples for to test datasette JS api
+ */
+const addPlugins = (manager) => {
+
+  manager.registerPlugin("column-name-plugin", {
+    version: 0.1,
+    getColumnHeaderItems: (columnMeta) => {
```
Comment (simonw) [inline datasette/static/table-example-plugins.js:18]: I'd call this getColumnActions() to match with the names of other server-side action plugins: https://docs.datasette.io/en/stable/plugin_hooks.html#table-actions-datasette-actor-database-table-request

**Comment by hydrosquall** [comment]
Notes from 1:1 - it _is_ possible to pass in URL params into a ObservableHQ notebook: https://observablehq.com/@bherbertlc/pass-values-as-url-parameters

**Comment by hydrosquall** [comment]
## Why

The Datasette JS Plugin API allows developers to add interactive features to the UI without having to modify the Python source code. 

## Setup

No external/NPM dependencies are needed.

Plugin behavior is coordinated by the Datasette `manager`. Every page has 1 `manager`.

There are 2 ways to add your plugin to the `manager`.

1. Read `window.__DATASETTE__` if the manager was already loaded.

```js
const manager = window.__DATASETTE__;
```

2. Wait for the `InitDatasette` event to fire if your code was loaded before the manager is ready. 

```js
document.addEventListener("InitDatasette", function (evt) {
  const { detail: manager } = evt;
  
  // register plugin here
});
```

3. Add plugin to the manager by calling `manager.registerPlugin`. Each plugin will supply 1 or more hooks with

- unique name (`YOUR_PLUGIN_NAME`)
- a numeric version (starting at `0.1`), 
- configuration value, the details vary by hook. (In this example, `getColumnActions` takes a function)

```js
manager.registerPlugin("YOUR_PLUGIN_NAME", {
    version: 0.1,
    getColumnActions: (columnMeta) => {
      return [
        {
          label: "Copy name to clipboard",
          // evt = native click event
          onClick: (evt) => copyToClipboard(columnMeta.column),
        }
      ];
    },
  });
```

There are 2 plugin hooks available to `manager.registerPlugin`:

1. `getColumnActions` - Add items to the cog menu for headers on datasette table pages
2. `renderAboveTablePanel` - Add items to "tabbed" panel above the `<table/>` on pages that use the Datasette table template.

While there are additional properties on the `manager`, but we're it's not advised to depend on ghtem directly as the shape is subject to change.

We welcome ideas for more hooks, or feedback on the current design!

## Examples

See the [example plugins file](https://github.com/simonw/datasette/blob/2d92b9328022d86505261bcdac419b6ed9cb2236/datasette/static/table-example-plugins.js) for additional examples.


## Hooks API Guide

### `renderAboveTablePanel`

Provide a function with a list of panel objects. Each panel object should contain

1. A unique string `id`
2. A string `label` for the tab
3. A `render` function. The first argument is reference an HTML Element.

Example:

```js
 manager.registerPlugin("panel-plugin-graphs", {
    version: 0.1,
    getAboveTablePanelConfigs: () => {
      return [
        {
          id: 'first-panel',
          label: "My new panel",
          render: node => {
            const description = document.createElement('p');
            description.innerText = 'Hello world';
            node.appendChild(description);
          }
        }
      ];
    },
  });
```

### `getColumnActions`

Provide a function that returns a list of action objects. Each action object has

1. A string `label` for the menu dropdown label
2. An onClick `render` function.


Example:

```js
  manager.registerPlugin("column-name-plugin", {
    version: 0.1,
    getColumnActions: (columnMeta) => {
      
      // Info about selected column. 
      const { column, columnNotNull,columnType, isPk } = columnMeta;

      return [
        {
          label: "Copy name to clipboard",
          onClick: (evt) => copyToClipboard(column),
        }
      ];
    },
  });
```

The getColumnActions callback has access to an object with metadata about the clicked column. These fields include:

- column: string (name of the column)
- columnNotNull: 0 or 1: 0 (false) or 1 (true)
- columnType: sqlite datatype enum (text, number, etc)
- isPk: Whether this is the primary key: 0 (false) or 1 (true)

You can use this column metadata to customize the action config objects (for example, handling different summaries for text vs number columns).

**Comment by hydrosquall** [comment]
## Research notes

- I stuck to the "minimal dependencies" ethos of datasette (no React, Typescript, JS linting, etc).
- Main threads on JS plugin development
  - Main: sketch of pluggy-inspired system: https://github.com/simonw/datasette/issues/983
  - Main: provide locations in Datasette HTML that are designed for multiple plugins to safely cooperate with each other (starting with the panel, but eventually could extend to "search boxes"): https://github.com/simonw/datasette/issues/1191
  - Main: HTML hooks for JS plugin authors: https://github.com/simonw/datasette/issues/987
- Prior threads on JS plugins in Datasette for future design directions
  - Idea: pass useful strings to JS plugins: https://github.com/simonw/datasette/issues/1565
  - Idea: help with plugin dependency loading: https://github.com/simonw/datasette/issues/1542 . (IMO - the plugin providing the dependency can emit an event once it's done.  Other plugins can listen for it, or ask the manager to inform them when the dependency is available). 
  - Idea: help plugins to manage state in shareable URLs (plugins shouldn't have to interact with the URL directly, should have some basic insulation from clobbering each others' keys): https://github.com/simonw/datasette/issues/1144
- Articles on plugins reviewed
  - https://css-tricks.com/designing-a-javascript-plugin-system/
- Plugin/Extension systems reviewed (mostly JS).
  - Yarn: https://yarnpkg.com/advanced/plugin-tutorial
  - Tappable https://github.com/webpack/tapable (used by Auto, webpack)
  - Pluggy: https://pluggy.readthedocs.io/en/stable/
  - VSCode: https://code.visualstudio.com/api/get-started/your-first-extension
  - Chrome: https://developer.chrome.com/docs/extensions/reference/
  - Figma/Figjam Widget: https://www.figma.com/widget-docs/
  - Datadog Apps: [Programming Model](https://github.com/DataDog/apps/blob/master/docs/en/programming-model.md)
  - Storybook: https://storybook.js.org/docs/react/addons/addons-api

**Review by asg017 — COMMENTED** [review]
Hey! This is awesome: I really dig the pure vanilla JS plugin system (including the `"InitDatasette"` custom event + `window.__DATASETTE__`). 

- Left some very nitty comments, not too important
- For the plugin methods `getColumnActions` and `getAboveTablePanelConfigs`: Maybe they should have more "action-y" names like `addColumnActions` and `addAboveTablePanelConfigs`? It seems like the `get*` prefix is there because they're lifecycle/hook names, but it seems odd to be to have a method called `get*()` that returns new items
- I know TypeScript can be a little controversial, but for APIs like this (well defined callbacks and return objects) I find types to be extremely helpful. I think we can distribute an optional flat single-file `types.d.ts` file like so:

```ts
interface ColumnMeta {
  column: string;
  columnNotNull: boolean;
  columnType: string;
  isPk: boolean;
}

interface ColumnAction {
  label: string;
  onClick: (event: MouseEvent) => void;
}

interface AboveTablePanelConfig {
  id: string;
  label: string;
  render: (node: HTMLElement) => void;
}

export interface Plugin {
  version: number;
  getColumnActions(columnMeta: ColumnMeta): ColumnAction[];
  getAboveTablePanelConfigs(): AboveTablePanelConfig[];
}
export interface Manager {
  plugins: Map<String, Plugin>;
  registerPlugin(name: string, plugin: Plugin): void;
}

```

And then, one could use these types with JSDoc in plain JS files like so:

```js
/**
 * @typedef {import('./types').Manager} Manager
 * @typedef {import('./types').Plugin} Plugin
 */

document.addEventListener("InitDatasette", function (evt) {
  const { detail: manager } = evt;
  addPlugins(manager);
});

/**
 * @param {Manager} manager
 */
function addPlugins(manager) {
  manager.registerPlugin("column-name-plugin", {
    version: 0.1,
    getColumnActions: (columnMeta) => { ... }
  });
}
```

Then your editor will have type suggestions for `manager`, `columnMeta`, etc.

Probably not necessary for this PR, but happy to contribute a `types.d.ts` when this is finalized!

- Maybe we can make it easier to do pure-js datasette plugins? For example the `example_js_manager_plugins.py` file is fairly small that just uses `extra_js_urls`. Maybe there can be CLI options like:

```
datasette fixtures.db --plugin-js path/to/table-example-plugins.js`
```

And then do the `PERMITTED_VIEWS` filtering in JS rather than Python. Or even autodetect JS files in `--plugins-dir/` and server those automatically (though I doubt that's backwards compatible)

File: datasette/static/datasette-manager.js [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+   * New DOM elements created each time the button is clicked so the data is not stale.
+   * Items
+   *  - must provide label (text)
+   *  - might provide href (string) or an onclick ((evt) => void)
+   *
+   * columnMeta is metadata stored on the column header (TH) as a DOMStringMap
+   * - column: string
+   * - columnNotNull: 0 or 1
+   * - columnType: sqlite datatype enum (text, number, etc)
+   * - isPk: 0 or 1
```
[the remark this answers is not in the document]
Comment (hydrosquall) [reply datasette/static/datasette-manager.js:65]: I think that makes sense, right now I'm passing through the data unparsed directly from the `DOMStringMap`. 

Probably more ergonomic to use these precise datatypes and more opinionated field names, great suggestions - I'll apply these soon.

File: datasette/static/datasette-manager.js [hunk 3]
```diff
@@ -0,0 +1,202 @@
+// Custom events for use with the native CustomEvent API
+const DATASETTE_EVENTS = {
+  INIT: "InitDatasette", // returns datasette manager instance in evt.detail
```
[the remark this answers is not in the document]
Comment (hydrosquall) [reply datasette/static/datasette-manager.js:3]: I'm flexible, I saw a bunch of opinions about this here:

https://stackoverflow.com/questions/19071579/javascript-dom-event-name-convention

It looks like it might be safe to err on the side of all-lowercase. `datasette_init` or `datasette.init` both have a reasonable look to them.

File: datasette/static/datasette-manager.js [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+   * In future, panels could be repeated. We omit that for now since so many plugins depend on
+   * shared URL state, so having multiple instances of plugin at same time is problematic.
+   * Currently, we never destroy any panels, we just hide them.
+   *
+   * TODO: nicer panel css, show panel selection state.
+   * TODO: does this hook need to take any arguments?
+   */
+  renderAboveTablePanel: () => {
+
+    const aboveTablePanel = document.querySelector(DOM_SELECTORS.aboveTablePanel);
```
[the remark this answers is not in the document]
Comment (hydrosquall) [reply datasette/static/datasette-manager.js:94]: Interesting- in those cases, do you think it still makes sense for the plugin to attempt to display (in which case there should be a fallback mount point, or the mount point should always show up regardless of whether there is data), or should a panel like this only show up for views that have some rows?

I think it's the first case, but wanted to check what you think.

**Comment by hydrosquall** [comment]
Thanks for the thoughtful review and generous examples @asg017 ! I'll make the changes you suggested soon. Bonus thoughts inlined below.

> comments

These were very much appreciated, it's important to a plugin system that details like this feel right! I'll address them in batch later in the week. 

> I know TypeScript can be a little controversial

FWIW I am in favor of doing Typescript - I just wanted to keep the initial set of files in this PR as simple as possible to review. Really appreciate you scaffolding this initial set of types + I think it would be a welcome addition to maintain a set of types.d.ts files. 

I'm entertaining the idea of writing the actual source code in Typescript as long as the compiled output is readable b/c it can be tricky to keep the types and plain JS files in sync. Curious if you have encountered projects that are good at preventing drift.

> Maybe they should have more "action-y" names

This is a great observation. I'm inclined towards something like `make*` or `build*` since to me `add*` make me think the thing the method is attached to is being mutated, but I agree that any of these may be clearer than the current `get*` setup. I'll go through and update these. 

> Maybe we can make it easier to do pure-js datasette plugins?

I really like this idea! It'll be easier to get contributors if they don't have to touch the python side at _all_. 

> And then do the PERMITTED_VIEWS filtering in JS rather than Python.

One cost of doing this is that pages that won't use the JS would still have to load the unused code (given that I'm not sending up anything complex like lazy loading). But hopefully the manager core size is close to negligible, and it won't be a big deal.

File: datasette/static/datasette-manager.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (hydrosquall) [reply datasette/static/datasette-manager.js:97]: Decided to always provide a hook so this won't be an issue

https://github.com/simonw/datasette/pull/2052/commits/06b4829035d5fc48b896d4afa5efdf750cf6d895

but also added a bit of JS safety in case the template structure changes in the future.

https://github.com/simonw/datasette/pull/2052/commits/cf504fe49207ec71e0f66340e86fc8d697b437ab

**Review by asg017 — COMMENTED** [review]
lgtm! Very eager to get this in, I want to add a hook for adding new features to the SQL editor with custom codemirror components

**Comment by cldellow** [comment]
Thanks for putting this together! I've been slammed with work/personal stuff so haven't been able to actually prototype anything with this. :(

tl;dr: I think this would be useful immediately as is. It might also be nice if the plugins could return `Promise`s.

The long version: I read the design notes and example plugin. I think I'd be able to use this in [datasette-ui-extras](https://github.com/cldellow/datasette-ui-extras) for my lazy-facets feature.

The lazy-facets feature tries to provide a snappier user experience. It does this by altering how suggested facets work.

First, at page render time:
(A) it lies to Datasette and claims that no columns support facets, this avoids the lengthy delays/timeouts that can happen if the dataset is large.
(B) there's a python plugin that implements the [extra_body_script](https://docs.datasette.io/en/stable/plugin_hooks.html#extra-body-script-template-database-table-columns-view-name-request-datasette) hook, to write out the list of column names for future use by JavaScript

Second, at page load time: there is some JavaScript that:
(C) makes AJAX requests to suggest facets for each column - it makes 1 request per column, using the data from (B)
(D) wires up the column menus to add Facet-by-this options for each facet

With the currently proposed plugin scheme, I think (D) could be moved into the plugin. I'd do the ajax requests, then register the plugin.

If the plugin scheme also supported promises, I think (B) and (C) could also be moved into the plugin.

Does that make sense? Sorry for the wall of text!

**Comment by cldellow** [comment]
Also, just checking - is this how I'd write bulletproof plugin registration code that is robust against the order in which the script tags load (eg if both my code and the Datasette code are loaded via a `<script async src='...'/>` tag)?

```js
if (window.__DATASETTE__)
  go(window.__DATASETTE__);
else
  document.addEventListener("datasette_init", (evt) => go(evt.detail));

function go(manager) {
  manager.registerPlugin(...)
}
```

I don't know if it'd make sense, but you could also consider the asynchronous queuing pattern that Google Analytics uses (see [this Stack Overflow post](https://stackoverflow.com/questions/6963779/whats-the-name-of-google-analytics-async-design-pattern-and-where-is-it-used) for more details):

```
__DATASETTE__ = __DATASETTE__ || [];
__DATASETTE.push(go);

function go(manager) {
  manager.registerPlugin(...);
}
```

**Comment by hydrosquall** [comment]
Hey @cldellow , thanks for the thoughtful feedback and describing the "lazy facets" feature! 

It sounds like the [postTask](https://developer.mozilla.org/en-US/docs/Web/API/Scheduler/postTask) API might be relevant for the types of network request scheduling you have in mind. 

Addressing your points inline below:

> It might also be nice if the plugins could return Promises.

Were you picturing that the whole plugin config object could be returned as a promise, or that the individual hooks (like `makeColumnActions` or `makeAboveTablePanelConfigs` supported returning a promise of arrays instead only returning plain arrays?

I think what you're describing can be achievable, but I want to make sure I do so in a way that addresses your need / keeps the complexity of the plugin core system at a level this is approachable . 

I have a hunch that what you're describing might be achievable without adding Promises to the API with something like

```
fetch('/api/with-custom-facets').then(myFacets => {
    // reusing the go() idiom
   go(manager, myFacets);
})
```

but I'd like to confirm if that's the case before investigating adding support.

> bulletproof plugin registration code that is robust against the order in which the script tags load

Yes, I think what you wrote looks right to me! While it looks a little bit verbose compared to the second example, I'm hoping we can mitigate the cost of that during this API incubation phase by making it an easy-to-copy paste code snippet.

I haven't heard of the GA queing pattern before, thanks for the example. I won't have time to implement of proof of concept in the next few weeks, but I took some time to think through the pros/cons to decide whether we may want to add this in a future release:

I can see that this approach brings advantages

- Plugin developers don't need to know the name of the datasette initialization event to start their plugin
- Pushing a function to an array probably is easier (definitely more concise) than adding a document event listener
- One less event listener sitting in memory

It also has some minor costs

- A malicious plugin could choose to (or accidentally) mess with the order of the queue if multiple scripts are lined up
- Some risk in encouraging people to mutate global state
- (not a cost, more a moot point): changing this API may not make a meaningful difference if we're discussing whether people enter 2 vs 5 lines of code, especially if those lines are encapsulated by a function we provide (maybe something that's available on the `window` provided by Datasette as an inline script tag).

**Comment by cldellow** [comment]
> Were you picturing that the whole plugin config object could be returned as a promise, or that the individual hooks (like makeColumnActions or makeAboveTablePanelConfigs supported returning a promise of arrays instead only returning plain arrays?

The latter - that you could return a promise of arrays, so it parallels the ["await me maybe" pattern in Datasette](https://simonwillison.net/2020/Sep/2/await-me-maybe/), where you can return either a value, a callable or an awaitable.

> I have a hunch that what you're describing might be achievable without adding Promises to the API with something

Oops, I did a poor job explaining. Yes, this would work - but it requires me to continue to communicate the column names out of band (in order to fetch the facet data per-column before registering my plugin), vs being able to re-use them from the plugin implementation.

This isn't that big of a deal - it'd be a nice ergonomic improvement, but nowhere near as a big of an improvement as having an officially sanctioned way to add stuff to the column menus in the first place.

This could also be layered on in a future commit without breaking v1 users, too, so it's not at all urgent.

> especially if those lines are encapsulated by a function we provide (maybe something that's available on the window provided by Datasette as an inline script tag

Ah, this is maybe the the key point. Since it's all hosted inside Datasette, Datasette can provide some arbitrary sugar to make it easier to work with.

My experience with async scripts in JS is that people sometimes don't understand the race conditions inherent to them. If they copy/paste from a tutorial, it does just work. But then they'll delete half the code, and by chance it still works on their machine/Datasette templates, and now someone's headed for an annoying debugging session -- maybe them, maybe someone else who tries to re-use their plugin.

Again, a fairly minor thing, though.

**Comment by hydrosquall** [comment]
Thanks @cldellow for the thoughtful comments! These are all things that I'll keep in mind as we figure out how/if this API is actually used by plugin authors once it's actually out in the world.

> Yes, this would work - but it requires me to continue to communicate the column names out of band (in order to fetch the facet data per-column before registering my plugin), vs being able to re-use them from the plugin implementation.

Ah, I understand now! Thanks for explaining.

**Comment by asg017** [comment]
:wave: would love to see this get merged soon! I want to make a javascript plugin on top of the code-mirror editor to make a few things nicer (function auto-complete, table/column descriptions, etc.), and this would help out a bunch

**Comment by asg017** [comment]
Ok @hydrosquall a few more things before this PR should be good to go:

- Can we move `datasette/static/table-example-plugins.js` into `demos/plugins/static`?
- For `datasetteManager.VERSION`, can we fill that in or just comment it out for now? Not sure how difficult it'll be to inject it server-side



- [``]()
- [``]()
- [``]()
- [``]()
- [``]()
- [``]()

**Comment by asg017** [comment]
Ok @hydrosquall a couple things before this PR should be good to go:

- Can we move `datasette/static/table-example-plugins.js` into `demos/plugins/static`?
- For `datasetteManager.VERSION`, can we fill that in or just comment it out for now? Not sure how difficult it'll be to inject it server-side. I imagine we could also have a small build process with esbuild/rollup that just injects a version string into `manager.js` directly, so we don't have to worry about server-rendering (but that can be a future PR)

In terms of how to integrate this into Datasette, a few options I can see working:

- Push this as-is and figure it out before the next release
- Hide this feature behind a settings flag (`--setting unstable-js-plugins on`) and use that setting to hide/show `<script src="{{ urls.static('datasette-manager.js') }}" defer></script>` in `base.html`

I'll let @simonw decide which one to work with. I kindof like the idea of having an "unstable" opt-in process to enable JS plugins, to give us time to try it out with a wide variety of plugins until we feel its ready.

I'm also curious to see how "plugins for a plugin' would work, like #1542. For example, if the leaflet plugin showed default markers, but also included its own hook for other plugins to add more markers/styling. I'm imagine that the individual plugin would re-create their own plugin system compared to this, since handling "plugins of plugins" at the top with Datasette seems really convoluted. 

Also for posterity, here's a list of Simon's Datasette plugins that use "extra_js_urls()", which probably means they can be ported/re-written to use this new plugin system:

- [`datasette-vega`](https://github.com/simonw/datasette-vega/blob/00de059ab1ef77394ba9f9547abfacf966c479c4/datasette_vega/__init__.py#L25)
- [`datasette-cluster-map`](https://github.com/simonw/datasette-cluster-map/blob/795d25ad9ff6cba0307191f44fecc8f8070bef5c/datasette_cluster_map/__init__.py#L14)
- [`datasette-leaflet-geojson`](https://github.com/simonw/datasette-leaflet-geojson/blob/64713aa497750400b9ac2c12e8bb6ffab8eb77f3/datasette_leaflet_geojson/__init__.py#L47)
- [`datasette-pretty-traces`](https://github.com/simonw/datasette-pretty-traces/blob/5219d65eca3d7d7a73bb9d3120df42fe046a1315/datasette_pretty_traces/__init__.py#L5)
- [`datasette-youtube-embed`](https://github.com/simonw/datasette-youtube-embed/blob/4b4a0d7e58ebe15f47e9baf68beb9908c1d899da/datasette_youtube_embed/__init__.py#L55)
- [`datasette-leaflet-freedraw`](https://github.com/simonw/datasette-leaflet-freedraw/blob/8f28c2c2080ec9d29f18386cc6a2573a1c8fbde7/datasette_leaflet_freedraw/__init__.py#L66)
- [`datasette-hovercards`](https://github.com/simonw/datasette-hovercards/blob/9439ba46b7140fb03223faff0d21aeba5615a287/datasette_hovercards/__init__.py#L5)
- [`datasette-mp3-audio`](https://github.com/simonw/datasette-mp3-audio/blob/4402168792f452a46ab7b488e40ec49cd4b12185/datasette_mp3_audio/__init__.py#L6)
- [`datasette-geojson-map`](https://github.com/simonw/datasette-geojson-map/blob/32af5f1fd1a07278bbf8071fbb20a61e0f613246/datasette_geojson_map/__init__.py#L30)

**Comment by hydrosquall** [comment]
> Ok @hydrosquall a couple things before this PR should be good to go:

Thank you @asg017 ! I've pushed both suggested changes onto this branch.

> Not sure how difficult it'll be to inject it server-side

If we are OK with having a build system, it would free me up to do do many things! We could make datasette-manager.js a server-side rendered file as a "template" instead of having it as a static JS file, but I'm not sure it's worth the extra jump in complexity / loss of syntax highlighting in the JS file.

In the short-term, I could see an intermediary solution where a unit test in the preferred language was able to read both `version.py` and `datasette-manager.js`, and make sure that the strings versions are in sync. (This assumes that we want the manager and datasette's versions to be synced, and not decoupled). Since the version is not changing very often, a "manual sync" might be good enough. 

> In terms of how to integrate this into Datasette, a few options I can see working:

This sounds good to me. I'm not sure how to add a settings flag, but will be interested to see the PR that adds support for it.

> I'm also curious to see how "plugins for a plugin' would work

I'm comfortable to wait until we have a realistic usecase for this. In the short term, I think we could give plugins a way to grant access to a "public API of other plugins", and also ask to be notified when plugins with other names have loaded, but don't picture the datasette manager getting more involved than that. 

> here's a list of Simon's Datasette plugins that use "extra_js_urls()"

Neat, thanks for compiling this list! Just curious, is there a query that can be used to compile this programmatically, or did you identify these through memory?

> I want to make a javascript plugin on top of the code-mirror editor to make a few things nicer (function auto-complete, table/column descriptions, etc.)

I look forward to trying this out 👍

**Comment by asg017** [comment]
> Just curious, is there a query that can be used to compile this programmatically, or did you identify these through memory?

I just did a github search for `user:simonw "def extra_js_urls("` ! Though I'm sure other plugins made by people other than Simon also exist out there https://github.com/search?q=user%3Asimonw+%22def+extra_js_urls%28%22&type=code

File: datasette/static/datasette-manager.js [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+  // multiple plugins can all request the same copy of leaflet.
+};
+
+
+const initializeDatasette = () => {
+  // Hide the global behind __ prefix. Ideally they should be listening for the
+  // DATASETTE_EVENTS.INIT event to avoid the habit of reading from the window.
+
+  window.__DATASETTE__ = datasetteManager;
+  console.debug("Datasette Manager Created!");
```
Comment (simonw) [inline datasette/static/datasette-manager.js:196]: I was worried that there might be browsers in which this would cause an error (because `console.debug` might not be defined), but as far as I can tell this has been supported in every modern browser for years at this point: https://console.spec.whatwg.org/ and https://developer.mozilla.org/en-US/docs/Web/API/console#browser_compatibility - so this is fine.

File: datasette/static/datasette-manager.js [hunk 6]
```diff
@@ -0,0 +1,211 @@
...
+};
+
+
+/**
+ * Monolith class for interacting with Datasette JS API
+ * Imported with DEFER, runs after main document parsed
+ * For now, manually synced with datasette/version.py
+ */
+const datasetteManager = {
+  VERSION: '1.0a2',
```
Comment (simonw) [inline datasette/static/datasette-manager.js:35]: I think we should make this available in the `base.html` template, something like this:
```html
<script>
datasetteVersion = '{{ datasette_version }}';
</script>
```
Then the `datasette-manager.js` script, which is loaded later, can pick it up from there.

The `{{ datasette_version }}` variable is already set and is used by the footer template here:

https://github.com/simonw/datasette/blob/0f7192b6154edb576c41b55bd3f2a3f53e5f436a/datasette/templates/_footer.html#L1

Comment (hydrosquall) [reply datasette/static/datasette-manager.js:35]: Brilliant! Confirmed that this appears to work: 

https://a.cl.ly/Z4uGrYg8

**Comment by hydrosquall** [comment]
Thanks for the review and the code pointers @simonw - I've made the suggested edits, fixed the renamed variable, and confirmed that the panels still render on the `table` and `database` views.

**Comment by 20after4** [comment]
This is such a well thought out contribution. I don't think I've seen such a thoroughly considered PR on any project in recent memory.

**Comment by simonw** [comment]
Weird, the `cog` check is failing in CI.

**Comment by simonw** [comment]
Oh! I think I broke Cog on `main` and these tests are running against this branch rebased against main.

**Comment by simonw** [comment]
I'm landing this despite the cog failures. I'll fix them on main if I have to.
